---
title: "TLimmuno2 outperforms existing methods in peptide-MHC II immunogenicity prediction"
author: "Guangshuai Wang"
date: "`r Sys.Date()`"
output:   
  rmdformats::readthedown:
    highlight: kate
    lightbox: false
    toc_depth: 3
    mathjax: true
---

```{r , include=FALSE}
options(max.print = "120")
knitr::opts_chunk$set(echo = TRUE, comment = "#>", eval = TRUE, collapse = TRUE,cache = FALSE,warning = FALSE)
knitr::opts_knit$set(width = 120)
```

```{r , include=FALSE}
library(tidyverse)
library(reticulate)
library(DT)
use_python(Sys.which("python"))
```

```{python, include = FALSE}
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
mpl.rcParams['pdf.fonttype'] = 42
mpl.rcParams['ps.fonttype'] = 42
plt.rcParams.update({'font.family':'Arial'})
import seaborn as sns
import numpy as np
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.metrics import roc_curve,auc,confusion_matrix,precision_score,recall_score,precision_recall_curve
import matplotlib as mpl
import sys
sys.path.append("../Python")
from Delongtest import DelongTest
```

This part focuses on another independent dataset to compare the performance of TLimmuno2. it contains the following sections:

- [Benchmark in immunoigenicity](#benchmark-in-immunoigenicity)
- [Structure immportance](#structure-immportance)
- [Neoepitope and wildtype peptide](#neoepitope-and-wildtype-peptide)
- [Position importance](#position-importance)

## Benchmark in immunoigenicity

The IEDB-Benchmark dataset is from the same database as the training set, thus they may have similar distribution patterns, which may lead to systematic bias. To perform an unbiased performance evaluation in predicting immunogenic peptides, we retrieved published papers to collect neoepitopes that were tested experimentally for CD4+ T cell immunogenicity.


### Data proprecessing

```{python}
#python
peptide = pd.read_csv("../data/nature_peptide.csv")
HLA = pd.read_csv("../data/nature_HLA.csv")
peptide = peptide.drop_duplicates()
peptide["Immunogenicity"].value_counts()
peptide["Len"] = peptide["Peptide"].map(len)
peptide[peptide["Len"]>=15]["Peptide"].to_csv("../data/pep_only.txt",header = None,index = None,sep = " ")#chain indexing
IEDB_pep = peptide[peptide["Len"]>=15].reset_index(drop = True)
IEDB_pep["num"] = [x+1 for x in range(len(IEDB_pep))]#used to predict by IEDB tools
pep_ID = ["Pep{}".format(x+1) for x in range(len(peptide))]
peptide["pep_ID"] = pep_ID
fig,ax = plt.subplots()
sns.countplot(data = peptide,x = "Len")
plt.show()
```

```{python}
#python
ax = sns.countplot(data=peptide, x="Immunogenicity")
#ax.bar_label(ax.containers[0])
ax.text(-0.1, 860, "{:.3f}".format(851/931))
ax.text(0.9, 90, "{:.3f}".format(80/931))
plt.savefig("../figure/nature_immuno_rate.pdf",dpi = 300,transparent=True)
plt.show()
```

Neoepitopes tested experimentally were usually sequences of 20â€“25 amino acids, so we use 15-mer and k-mer methods (Methods) to assign a rank value to a given neoepitope.

For the 15mer method, we split one neoepitope sequence into 15 subsequences and score or rank each subsequence. In the kmer method, neoepitopes were cut into 13-21-mers subsequence. The prediction of the neoepitope is the maximum score or the minimum ranking of the predicted results of all subsequences by different methods. In case of a neoepitope with multiple MHC types, we used the maximum score or minimum rank of prediction for different MHC types as the prediction result for the neoepitope.

```{python}
#python
Data = pd.merge(HLA,peptide)
IMM_train_data_allele = pd.read_csv("../data/IEDB_MHCII_immuno_allele.csv",header = None,names = ["HLA"])
Data = Data[Data["HLA"].isin(IMM_train_data_allele["HLA"])]
pseudo_seq = pd.read_table("../data/pseudosequence.2016.all.X.dat",header = None,names = ["HLA","sequence"])
Data1 = pd.merge(Data,pseudo_seq)
Data1["length"] = Data1["Peptide"].map(len)
Data1 = Data1[Data1["length"]>=13]
Data1 = Data1.reset_index(drop = True)
#15mer
def mer15(Pep):
    P = []
    Length = 15
    for i in range(len(Pep) - Length +1):
        pep = Pep[i:i+Length]
        P.append(pep)
    return P
DF15 = pd.DataFrame()
for i in range(len(Data1)):
    pep = mer15(Data1["Peptide"][i])
    df = pd.DataFrame(pep)
    df["pep_ID"] = Data1["pep_ID"][i]
    df["HLA"] = Data1["HLA"][i]
    DF15 = pd.concat([DF15,df])
DF15.columns = ["pep","pep_ID","HLA"]
benchmark_data = pd.merge(DF15,pseudo_seq)
benchmark_data.to_csv("../data/15mer_data.csv")
benchmark_data["pep"].to_csv("../data/15mer_pep_only.txt",header = None,index = None,sep = " ")
#kmer
def kmer(Pep,sa,end):
    P = []
    R = range(sa,end+1)
    for i in R:
        Length = i
        for i in range(len(Pep) - Length +1):
            pep = Pep[i:i+Length]
            P.append(pep)
    return P
DF_k = pd.DataFrame()
for i in range(len(Data1)):
    pep = kmer(Data1["Peptide"][i],13,21)
    df = pd.DataFrame(pep)
    df["pep_ID"] = Data1["pep_ID"][i]
    df["HLA"] = Data1["HLA"][i]
    DF_k = pd.concat([DF_k,df])
DF_k.columns = ["pep","pep_ID","HLA"]
benchmark_data = pd.merge(DF_k,pseudo_seq)
benchmark_data.to_csv("../data/kmer_data.csv")
```

### Different tools

We also compared the performance of TLimmuno2 with other immunogenicity prediction tools, including IEDB tool (CD4episcore) and Repitope. Note some published methods, such as Deepitope and FIONA do not have the tools publicly available, and are thus not included in the subsequent comparison analysis. we can't link their website for FIONA and we can't find the Github repositories for Deepitope.

#### IEDB online tool(CD4episcore)

We run IEDB in their website [IEDB online tool](http://tools.iedb.org/CD4episcore/), and this tool only support 15mer peptides.

#### Repitope

You can find the [Repitope](https://github.com/masato-ogishi/Repitope) in their Github, This tool is very slow and potentially problematic on larger datasets, so we only used 15mer peptides.

```{r,eval=FALSE}
#PBS
#R
library(tidyverse)
library(data.table)
library(Repitope)
options(java.parameters="-Xmx60G")

fragLibDT <- fst::read_fst("~/Project/MHCII/Repitope/FragmentLibrary_TCRSet_Public_RepitopeV3.fst", as.data.table=T)
featureDF_MHCII <- fst::read_fst("~/Project/MHCII/Repitope/FeatureDF_MHCII_Weighted.10000_RepitopeV3.fst", as.data.table=T)

data <- read_csv("~/Project/MHCII/Repitope/15mer_data.csv")
peptideSet_of_interest <- data$pep
peptideSet=peptideSet_of_interest
len = sapply(data$pep, nchar)
table(len)

res_MHCII <- EpitopePrioritization(
  featureDF=featureDF_MHCII[Peptide%in%MHCII_Human$Peptide,], 
  metadataDF=MHCII_Human[,.(Peptide,Immunogenicity)],
  peptideSet=peptideSet_of_interest,
  peptideLengthSet=11:30,
  fragLib=fragLibDT,
  aaIndexIDSet="all",
  fragLenSet=3:11,
  fragDepth=10000,
  fragLibType="Weighted",
  featureSet=MHCII_Human_MinimumFeatureSet,
  seedSet=1:1,
  coreN=1,
  outDir="~/Project/MHCII/Repitope/pre_result"  ## Intermediate and final output files will be stored under this directory
)
```

#### netMHCIIpan BA&EL

This part code is similar to the IEDB-benchmark dataset, only change the input and output file, so we don't add code in this part.

#### TLimmuno2

This part code is similar to the IEDB-benchmark dataset, only change the input and output file, so we don't add code in this part.

### Result

```{python 15mer}
#python
def IMM_process(Data,peptide):
    sam_pep_ID = Data["pep_ID"].unique()
    res = []
    p_ID = []
    Data["length"] = Data["pep"].map(len)
    for a in sam_pep_ID:
        Mean = []    
        ID = Data[Data["pep_ID"] == a]
        for i in ID["HLA"].unique():
            Data_HLA = ID[ID["HLA"] == i]
            prediction = []
            for l in Data_HLA["length"].unique():
                Data_len = Data_HLA[Data_HLA["length"] == l]
                pre = Data_len["prediction"].max()
                prediction.append(pre)
            x = np.max(prediction)
            Mean.append(x)
        res.append(max(Mean))
        p_ID.append(a)
    result = pd.DataFrame({"pep_ID":p_ID,"prediction":res})
    result = result.merge(peptide,how = "inner",on = "pep_ID" )
    return result
  
def auc_pr(data,true_label,prediction,rank = False):
    if rank:
        fpr,tpr,_ = roc_curve(data[true_label],1-data[prediction])
        precision,recall,_ = precision_recall_curve(data[true_label],1-data[prediction])
    else:
        fpr,tpr,_ = roc_curve(data[true_label],data[prediction])
        precision,recall,_ = precision_recall_curve(data[true_label],data[prediction])
        
    AUC = auc(fpr,tpr)
    PR = auc(recall,precision)
    
    return fpr,tpr,recall,precision,AUC,PR
```

```{python}
result = pd.read_csv("../data/15mer_result.csv")
result_15mer = IMM_process(result,peptide)
mer15_result = auc_pr(result_15mer,"Immunogenicity","prediction")
print("AUC:{}".format(mer15_result[4]))
print("PR_AUC:{}".format(mer15_result[5]))
```

```{python kmer}
#python
result = pd.read_csv("../data/kmer_result.csv")
result_kmer = IMM_process(result,peptide)
result_kmer = result_kmer[result_kmer["pep_ID"].isin(result_15mer["pep_ID"])]
kmer_result = auc_pr(result_kmer,"Immunogenicity","prediction")
print("AUC:{}".format(kmer_result[4]))
print("PR_AUC:{}".format(kmer_result[5]))
```

```{python IEDB_online_tool}
#python
IEDB_res = pd.read_csv("../data/IEDB_CD4_tools_manually.csv")
Imm_score = []
pro_num = []
for i in IEDB_res["Protein Number"].unique():
    y = IEDB_res[IEDB_res["Protein Number"] == i]
    imm_score = y["Immunogenicity Score"].mean()
    pro_num.append(i)
    Imm_score.append(imm_score)
IEDB_result = pd.DataFrame({"num":pro_num,"score":Imm_score})
IEDB_result = pd.merge(IEDB_result,IEDB_pep)
IEDB_result = IEDB_result[IEDB_result["Peptide"].isin(result_15mer["Peptide"])]
IEDB_score = auc_pr(IEDB_result,"Immunogenicity","score",rank = True)
print("AUC:{}".format(IEDB_score[4]))
print("PR_AUC:{}".format(IEDB_score[5]))
```


```{python Repitope}
#python
Repitope_result = pd.read_csv("../data/repitope_15mer_result.csv")
Repitope = pd.merge(Repitope_result,DF15,left_on = "Peptide",right_on = "pep")
sam_pep_ID = Repitope["pep_ID"].unique()
res = []
p_ID = []
for a in sam_pep_ID:
    Mean = []
    ID = Repitope[Repitope["pep_ID"] == a]
    for i in ID["HLA"].unique():
        x = ID[ID["HLA"] == i]
        x_mean = x["ImmunogenicityScore"].mean()
        Mean.append(x_mean)
    res.append(max(Mean))
    p_ID.append(a)
dic = {"pep_ID":p_ID,"ImmunogenicityScore":res}
Repitope_pre = pd.DataFrame(dic)
Repitope_result = Repitope_pre.merge(peptide,how = "inner",on = "pep_ID" )
Repitope_score = auc_pr(Repitope_result,"Immunogenicity","ImmunogenicityScore")
print("AUC:{}".format(Repitope_score[4]))
print("PR_AUC:{}".format(Repitope_score[5]))
```

```{python netMHCIIpan}
#python
def netMHCIIpan_process(result,allele,ori_sorted,peptide):
    
    #combine two type allele
    result = pd.merge(result,allele,left_on="allele",right_on="Allele Name")
    result_sorted = result.sort_values(by=["peptide","HLA"]).reset_index(drop = True)
    #combine with ori data
    result_combined = pd.merge(result_sorted,ori_sorted[["pep","pep_ID"]],left_on = "peptide",right_on ="pep")
    #result_combined["pep_length"] = result_combined["peptide"].map(len)
    sam_pep_ID = result_combined["pep_ID"].unique()
    res = []
    p_ID = []
    for a in sam_pep_ID:
        Mean = []
        ID = result_combined[result_combined["pep_ID"] == a]
        #get Mean of all splited pep and Max of all HLA
        for i in ID["Allele Name"].unique():
            Data_HLA = ID[ID["Allele Name"] == i]
            prediction = []
            for l in Data_HLA["pep_length"].unique():
                Data_len = Data_HLA[Data_HLA["pep_length"] == l]
                pre = Data_len["percentile_rank"].min()
                prediction.append(pre)
            x = np.min(prediction)
        res.append(np.min(x))
        p_ID.append(a) 
    final_result = pd.DataFrame({"pep_ID":p_ID,"prediction":res})
    final_result = final_result.merge(peptide,how = "inner",on = "pep_ID" )
    return final_result 
```

```{python}
DF15_sorted = DF15.sort_values(by=["pep","HLA"]).reset_index(drop = True)
allele = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/allele.csv")
netMHCIIpan_ba = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/netMHCIIpan_15_mer_ba.csv")
netMHCIIpan_ba["pep_length"] = netMHCIIpan_ba["peptide"].map(len)
net_15mer_ba = netMHCIIpan_process(netMHCIIpan_ba,allele,DF15_sorted,peptide)
mer15_ba_result = auc_pr(net_15mer_ba,"Immunogenicity","prediction",rank = True)
print("AUC:{}".format(mer15_ba_result[4]))
print("PR_AUC:{}".format(mer15_ba_result[5]))
```

```{python}
netMHCIIpan_el = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/netMHCIIpan_15_mer_el.csv")
netMHCIIpan_el["pep_length"] = netMHCIIpan_el["peptide"].map(len)
net_15mer_el = netMHCIIpan_process(netMHCIIpan_el,allele,DF15_sorted,peptide)
mer15_el_result = auc_pr(net_15mer_el,"Immunogenicity","prediction",rank = True)
print("AUC:{}".format(mer15_el_result[4]))
print("PR_AUC:{}".format(mer15_el_result[5]))
```

```{python}
DF_k_sorted = DF_k.sort_values(by=["pep","HLA"]).reset_index(drop = True)
netMHCIIpan_ba_k = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/netMHCIIpan_k_mer_ba.csv")
netMHCIIpan_ba_k["pep_length"] =netMHCIIpan_ba_k["peptide"].map(len)
net_kmer_ba = netMHCIIpan_process(netMHCIIpan_ba_k,allele,DF_k_sorted,peptide)
kmer_ba_result = auc_pr(net_kmer_ba,"Immunogenicity","prediction",rank = True)
netMHCIIpan_el_k = pd.read_csv("../data/netMHCIIpan_k_mer_el.csv")
netMHCIIpan_el_k["pep_length"] =netMHCIIpan_el_k["peptide"].map(len)
net_kmer_el = netMHCIIpan_process(netMHCIIpan_el_k,allele,DF_k_sorted,peptide)
kmer_el_result = auc_pr(net_kmer_el,"Immunogenicity","prediction",rank = True)
print("BA AUC: {}".format(kmer_ba_result[4]))
print("BA PR_AUC: {}".format(kmer_ba_result[5]))
print("EL AUC: {}".format(kmer_el_result[4]))
print("EL PR_AUC: {}".format(kmer_el_result[5]))
```

```{python}
kmer_data = pd.read_csv("../data/kmer_data.csv")
kmer_data["pep"].to_csv("../data/kmer_pep_only.csv",header = None,index = None)
```

In these comparisons, TLimmuno2 shows the best performance both in the 15-mer method and the k-mer method. Repitope and IEDB tools evaluate epitope immunogenicity using population MHC information, not considering the personalized MHC allele information, and have a poor performance in these comparisons.

```{python 15mer_ROC_fig}
#python
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
ax.plot(mer15_result[0],mer15_result[1],label='TLimmuno2 : (AUC={0:.4f})'.format(mer15_result[4]), lw=lw)
ax.plot(IEDB_score[0],IEDB_score[1],label='IEDB : (AUC={0:.4f})'.format(IEDB_score[4]), lw=lw)
ax.plot(Repitope_score[0],Repitope_score[1],label='Repitope : (AUC={0:.4f})'.format(Repitope_score[4]), lw=lw)
ax.plot(mer15_ba_result[0],mer15_ba_result[1],label='NetMHCIIpan_ba : (AUC={0:.4f})'.format(mer15_ba_result[4]), lw=lw)
ax.plot(mer15_el_result[0],mer15_el_result[1],label='NetMHCIIpan_el : (AUC={0:.4f})'.format(mer15_el_result[4]), lw=lw)
ax.set_ylim([0,1.05])
ax.set_xlim([0,1])
ax.set_xlabel('False Positive Rate',fontsize = 14)
ax.set_ylabel('True Positive Rate',fontsize = 14)
ax.legend(loc="lower right",fontsize = 12)
plt.savefig("../figure/benchmark_15mer.pdf",dpi = 300,transparent=True)
plt.show()
```

```{python kmer_ROC_fig}
#python
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
ax.plot(kmer_result[0],kmer_result[1],label='TLimmuno2 : (AUC={0:.4f})'.format(kmer_result[4]))
ax.plot(kmer_ba_result[0],kmer_ba_result[1],label='NetMHCIIpan_ba : (AUC={0:.4f})'.format(kmer_ba_result[4]))
ax.plot(kmer_el_result[0],kmer_el_result[1],label='NetMHCIIpan_el : (AUC={0:.4f})'.format(kmer_el_result[4]))
ax.set_ylim([0,1.05])
ax.set_xlim([0,1])
ax.set_xlabel('False Positive Rate',fontsize = 14)
ax.set_ylabel('True Positive Rate',fontsize = 14)
ax.legend(loc="lower right",fontsize = 12)
plt.savefig("../figure/benchmark_kmer.pdf",dpi = 300,transparent=True)
plt.show()
```

```{python 15mer_PR_fig}
#python
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot(mer15_result[2],mer15_result[3],label='TIMM2pred : (AUC={0:.4f})'.format(mer15_result[5]))
ax.plot(IEDB_score[2],IEDB_score[3],label='IEDB : (AUC={0:.4f})'.format(IEDB_score[5]))
ax.plot(Repitope_score[2],Repitope_score[3],label='Repitope : (AUC={0:.4f})'.format(Repitope_score[5]))
ax.plot(mer15_ba_result[2],mer15_ba_result[3],label='NetMHCIIpan_ba : (AUC={0:.4f})'.format(mer15_ba_result[5]))
ax.plot(mer15_el_result[2],mer15_el_result[3],label='NetMHCIIpan_el : (AUC={0:.4f})'.format(mer15_el_result[5]))

plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('Recall',fontsize = 14)
plt.ylabel('Precision',fontsize = 14)
plt.legend(loc="upper right",fontsize = 12)
plt.savefig("../figure/15mer_PR.pdf",dpi = 300,transparent=True)
plt.show()
```

```{python kmer_PR_fig}
#python
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot(kmer_result[2],kmer_result[3],label='TIMM2pred : (AUC={0:.4f})'.format(kmer_result[5]))
ax.plot(kmer_ba_result[2],kmer_ba_result[3],label='NetMHCIIpan_ba : (AUC={0:.4f})'.format(kmer_ba_result[5]))
ax.plot(kmer_el_result[2],kmer_el_result[3],label='NetMHCIIpan_el : (AUC={0:.4f})'.format(kmer_el_result[5]))
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('Recall',fontsize = 14)
plt.ylabel('Precision',fontsize = 14)
plt.legend(loc="upper right",fontsize = 12)
plt.savefig("../figure/kmer_pr.pdf",dpi = 300,transparent=True)
plt.show()
```

### Confusion matrix

```{python}
result_15mer["pre_label"] = result_15mer["prediction"].map(lambda x : 1 if x>0.5 else 0)
result_15mer_CM = confusion_matrix(result_15mer["Immunogenicity"],result_15mer["pre_label"])
result_15mer_CM = result_15mer_CM/len(result_15mer)
result_kmer["pre_label"] = result_kmer["prediction"].map(lambda x : 1 if x>0.5 else 0)
result_kmer_CM = confusion_matrix(result_kmer["Immunogenicity"],result_kmer["pre_label"])
result_kmer_CM = result_kmer_CM/len(result_kmer)
IEDB_result["pre_label"] = IEDB_result["score"].map(lambda x : 1 if x>50 else 0)
IEDB_result_CM = confusion_matrix(IEDB_result["Immunogenicity"],IEDB_result["pre_label"])
IEDB_result_CM = IEDB_result_CM/len(IEDB_result)
Repitope_result["pre_label"] = Repitope_result["ImmunogenicityScore"].map(lambda x : 1 if x>0.5 else 0)
Repitope_result_CM = confusion_matrix(Repitope_result["Immunogenicity"],Repitope_result["pre_label"])
Repitope_result_CM = Repitope_result_CM/len(Repitope_result)

def net_CM(data):
  data["pre_label"] = data["prediction"].map(lambda x : 1 if x<10 else 0)
  data_CM = confusion_matrix(data["Immunogenicity"],data["pre_label"])
  data_CM = data_CM/len(data)
  return data_CM


net_15mer_ba_CM = net_CM(net_15mer_ba)
net_15mer_el_CM = net_CM(net_15mer_el)
net_kmer_ba_CM = net_CM(net_kmer_ba)
net_kmer_el_CM = net_CM(net_kmer_el)
CM_all = [result_15mer_CM,result_kmer_CM,IEDB_result_CM,Repitope_result_CM,net_15mer_ba_CM,net_15mer_el_CM,net_kmer_ba_CM,net_kmer_el_CM]
```

```{python}
CM_all = [result_15mer_CM,IEDB_result_CM,Repitope_result_CM,net_15mer_ba_CM,net_15mer_el_CM,result_kmer_CM,net_kmer_ba_CM,net_kmer_el_CM]
CM_label = ["TLimmuno2_15mer","IEDB","Repitope","NetMHCIIpanBA_15mer","NetMHCIIpanEL_15mer","TLimmuon2_kmer","NetMHCIIpanBA_kmer","NetMHCIIpanEL_kmer"]
def CM_plot(data,label):
  fig,ax = plt.subplots(figsize = (4,4))
  sns.heatmap(data,annot =True,cmap="YlGnBu",vmin=0, vmax=1,cbar = False)
  plt.ylabel("True label")
  plt.xlabel("Prediction label")
  plt.title("{}".format(label))
  
  plt.savefig("../figure/{}.pdf".format(label),dpi = 300,transparent=True)

for i,l in enumerate(CM_all):
  CM_plot(l,CM_label[i])
  
```

## Structure immportance

Similarly, we want to demonstrate the impact of transfer learning on TLimmuno2, so we reconstructed two models, only BA and without BA. For the only BA model, its input layer contains only the transfer output of the BA model, while for the without BA, we masked the BA model and only the LSTM layer left. The detail code of this part is not inclued in the report. In independent validation datasets, the performance of these two models is significant decreased compared with TLimmuno2. These results demonstrated that the transfer learning approach improves the predictive power of TLimmuno2 in peptide-MHC II immunogenicity prediction.

```{python}
kmer_no_ba = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/kmer_no_ba_result.csv")
sam_pep_ID = kmer_no_ba["pep_ID"].unique()
sam_pep_ID = kmer_no_ba["pep_ID"].unique()
res = []
p_ID = []
for a in sam_pep_ID:
    Mean = []
    ID = kmer_no_ba[kmer_no_ba["pep_ID"] == a]
    for i in ID["HLA"].unique():
        x = ID[ID["HLA"] == i]
        #x_mean = x["Rank"].mean()
        x_mean = x["prediction"].max()
        Mean.append(x_mean)
    res.append(max(Mean))
    p_ID.append(a)
dic = {"pep_ID":p_ID,"prediction":res}
kmer_no_ba_result = pd.DataFrame(dic)
kmer_no_ba_result = kmer_no_ba_result.merge(peptide,how = "inner",on = "pep_ID" )
kmer_no_ba = auc_pr(kmer_no_ba_result,"Immunogenicity","prediction")
print("AUC:{}".format(kmer_no_ba[4]))
print("PR_AUC:{}".format(kmer_no_ba[5]))
kmer_only_ba = pd.read_csv("/Users/wangguangshuai/Data/MHCII/Data/IMM_model_benchmark/kmer_onlyBA_result.csv")
sam_pep_ID = kmer_only_ba["pep_ID"].unique()
res = []
p_ID = []
for a in sam_pep_ID:
    Mean = []
    ID = kmer_only_ba[kmer_only_ba["pep_ID"] == a]
    for i in ID["HLA"].unique():
        x = ID[ID["HLA"] == i]
        #x_mean = x["Rank"].mean()
        x_mean = x["prediction"].max()
        Mean.append(x_mean)
    res.append(max(Mean))
    p_ID.append(a)
dic = {"pep_ID":p_ID,"prediction":res}
kmer_only_ba_result = pd.DataFrame(dic)
kmer_only_ba_result = kmer_only_ba_result.merge(peptide,how = "inner",on = "pep_ID" )
kmer_only_ba = auc_pr(kmer_only_ba_result,"Immunogenicity","prediction")
print("AUC:{}".format(kmer_only_ba[4]))
print("PR_AUC:{}".format(kmer_only_ba[5]))
z_score_noba,p_val_noba = DelongTest(result_kmer["prediction"],kmer_no_ba_result["prediction"],result_kmer["Immunogenicity"]).show_result()
z_score_only_ba,p_val_only_ba = DelongTest(result_kmer["prediction"],kmer_only_ba_result["prediction"],result_kmer["Immunogenicity"]).show_result()
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
ax.plot(kmer_result[0],kmer_result[1],label='TLimmuno2 : (AUC={0:.4f})'.format(kmer_result[4]))
ax.plot(kmer_no_ba[0],kmer_no_ba[1],label='TLimmuno2_noba : (AUC={0:.4f})'.format(kmer_no_ba[4]))
ax.set_ylim([0,1.05])
ax.set_xlim([0,1])
ax.set_xlabel('False Positive Rate',fontsize = 14)
ax.set_ylabel('True Positive Rate',fontsize = 14)
ax.text(0.78,0.2,"P value: {:.4f}".format(p_val_noba),weight="bold")
ax.legend(loc="lower right",fontsize = 12)
plt.savefig("../figure/kmer_noba.pdf",dpi = 300,transparent=True)
plt.show()
fig,ax = plt.subplots(figsize = (8,8))
lw = 2
ax.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
ax.plot(kmer_result[0],kmer_result[1],label='TLimmuno2 : (AUC={0:.4f})'.format(kmer_result[4]))
ax.plot(kmer_only_ba[0],kmer_only_ba[1],label='TLimmuno2_onlyba : (AUC={0:.4f})'.format(kmer_only_ba[4]))
ax.set_ylim([0,1.05])
ax.set_xlim([0,1])
ax.set_xlabel('False Positive Rate',fontsize = 14)
ax.set_ylabel('True Positive Rate',fontsize = 14)
ax.text(0.78,0.2,"P value: {:.4f}".format(p_val_only_ba),weight="bold")
ax.legend(loc="lower right",fontsize = 12)
plt.savefig("../figure/kmer_onlyba.pdf",dpi = 300,transparent=True)
plt.show()
```


## Neoepitope and wildtype peptide 

To further demonstrate the discriminative power of TLimmuno2, we collected 52 MHC II neoepitopes and their corresponding wild-type peptides from published papers . These neoepitopes have been experimentally shown to elicit T cell responses. For each neoantigen and its corresponding wild-type peptide, we predicted the immunogenicity rank values by TLimmuno2. As shown in Figure below, the TLimmuno2 predicted immunogenicity rank values of mutant neopeptides are significantly lower than those of wild-type peptides, and this demonstrates the strong ability of TLimmuno2 in recognizing real neoepitope, even if the neoepitope differs from its wild-type peptide by only one amino acid.

```{python neoantigen}
#python
neoantigen = pd.read_csv("../data/neoantigen.csv")
pseudo = pd.read_table("../data/pseudosequence.2016.all.X.dat",header = None,names = ["allele","sequence"])
neoantigen_process = pd.merge(neoantigen,pseudo,left_on = "HLA",right_on = "allele").dropna()
neoantigen_process.to_csv("../data/neoantigen_pro.csv")
```

```{python neoantigen_res}
#python

result = pd.read_csv("../data/neoantigen_result.csv")
result = result[result["immunogenic"] == 1]
Rank = pd.melt(result[["WT_Rank","Mut_Rank"]],var_name = "Type",value_name = "Rank")
from scipy import stats
_,p = stats.wilcoxon(result["WT_Rank"],result["Mut_Rank"])
fig,ax = plt.subplots(figsize = (4,6))
ax = sns.boxplot(data = Rank,x = "Type",y = "Rank",width = 0.4,color = "skyblue",linewidth = 2,showfliers = False)
b = sns.stripplot(data = Rank,x = "Type",y = "Rank",color = "red")
x1, x2 = 0, 1   
y, h, col = Rank['Rank'].max() + 0.1, 0.1, 'k'
plt.plot([x1, x1, x2, x2], [y, y+h, y+h, y], lw=1, c=col)
plt.text((x1+x2)*.5, y+h, "**", ha='center', va='bottom', color=col)
ax.text(0.3,0.9,"P = {:.4f}".format(p),weight="bold")
plt.ylim([0,1.2])
plt.yticks(np.arange(0,1.2,0.2))
plt.xticks((0,1),labels = ["WT peptide","Neoantigen"])
plt.tight_layout()
sns.despine(offset = 3, trim = True)
plt.savefig("../figure/neoantigen.pdf",dpi = 300,transparent=True)
plt.show()
```
```{python}
#python
result = pd.read_csv("../data/neoantigen_result.csv")
result = result[result["immunogenic"] == 1]
Rank = pd.melt(result[["WT_prediction","Mut_prediction"]],var_name = "Type",value_name = "Score")
from scipy import stats
_,p = stats.wilcoxon(result["WT_prediction"],result["Mut_prediction"])
fig,ax = plt.subplots(figsize = (4,6))
ax = sns.boxplot(data = Rank,x = "Type",y = "Score",width = 0.4,color = "skyblue",linewidth = 2,showfliers = False)
b = sns.stripplot(data = Rank,x = "Type",y = "Score",color = "red")
x1, x2 = 0, 1   
y, h, col = Rank['Score'].max() + 0.1, 0.1, 'k'
plt.plot([x1, x1, x2, x2], [y, y+h, y+h, y], lw=1, c=col)
plt.text((x1+x2)*.5, y+h, "**", ha='center', va='bottom', color=col)
ax.text(0.3,0.9,"P = {:.4f}".format(p),weight="bold")
plt.ylim([0,1.2])
plt.yticks(np.arange(0,1.2,0.2))
plt.xticks((0,1),labels = ["WT peptide","Neoantigen"])
plt.tight_layout()
sns.despine(offset = 3, trim = True)
plt.savefig("../figure/neoantigen_prediction.pdf",dpi = 300,transparent=True)
plt.show()
```

## Position importance

### TLimmuno2

We performed in silico mutational analyses to investigate whether TLimmuno2 was able to learn the molecular characteristics of the peptide. By changing the amino acid residues at each position, we compare the differences of the predicted values, and the largest decrease corresponds to the greater importance of the position. We do this in two methods: ala-scanning and zero-setting. For ala-scanning, we sequentially replaced each position as the alanine; for zero-setting, we sequentially mutated each position (the representation vector for the mutated position is all zeros). We simulated this process 100 times and an ascending ranking was performed each time to highlight the most salient positions. 

```{python, eval = FALSE}
#python
#PBS
import pandas as pd
import tensorflow as tf
import sys
sys.path.append("/public/slst/home/wanggsh/Project/MHCII/utils")
from Blosum62 import blosum62
import numpy as np
import random
import matplotlib.pyplot as plt
import matplotlib as mpl
import os
#os.environ["CUDA_VISIBLE_DEVICES"]="-1" 

immuno_data = pd.read_csv("/public/slst/home/wanggsh/Data/MHCII_immuno/IMM_data_random2.csv")
immuno_data = immuno_data[immuno_data["label"] == 1]
immuno_data["length"] = immuno_data["Description"].map(len)
#choose 15mer
immuno_data = immuno_data[immuno_data["length"] == 15]
immuno_data = immuno_data.reset_index(drop=True)

immuno_data["pep_blosum"] = immuno_data["Description"].apply(blosum62,args=(21,))
immuno_data["MHC_blosum"] = immuno_data["sequence"].apply(blosum62,args=(34,))
peptide = np.empty((len(immuno_data),21,21))
for i in range(len(immuno_data)):
    peptide[i] = immuno_data["pep_blosum"][i].reshape((21,21))
#mhc 
MHC = np.empty((len(immuno_data),34,21))
for i in range(len(immuno_data)):
    MHC[i] = immuno_data["MHC_blosum"][i].reshape((34,21))
Ala = list(blosum62("A",1))

BAmodel = tf.keras.models.load_model("/public/slst/home/wanggsh/Project/Saved_model/new_BA_model/model1")
BAmodel = tf.keras.models.Model(inputs = BAmodel.input,outputs = BAmodel.layers[-2].output)
IMM = tf.keras.models.load_model("/public/slst/home/wanggsh/Project/Saved_model/MHCII_immuno/model")

n = 100
position_rank = np.empty([n,15])
for m in range(n):
    index = np.random.choice(np.arange(len(immuno_data)),2000)
    peptide_sam = peptide[index].copy()
    MHC_sam = MHC[index]
    BA = BAmodel.predict([peptide_sam,MHC_sam])
    pred_ori = IMM([peptide_sam,MHC_sam,BA]).numpy().mean()  
    array = []  
    for i in range(15):
        peptide_sam[:,i,:] = 0
        BA = BAmodel.predict([peptide_sam,MHC_sam])
        importance = pred_ori - IMM([peptide_sam,MHC_sam,BA]).numpy().mean()
        array.append(importance)
        peptide_sam = peptide[index].copy()
    ranking = np.argsort(array) + 1
    tmp = []
    for i in range(15):
        tmp.append(list(ranking).index(i+1))
    position_rank[m,:] = tmp
np.save("/public/slst/home/wanggsh/Data/MHCII_immuno/position_importance/position.npy",position_rank)

n = 100
position_rank_ala = np.empty([n,15])
for m in range(n):
    index = np.random.choice(np.arange(len(immuno_data)),2000)
    peptide_sam = peptide[index].copy()
    MHC_sam = MHC[index]
    BA = BAmodel.predict([peptide_sam,MHC_sam])
    pred_ori = IMM([peptide_sam,MHC_sam,BA]).numpy().mean()  
    array = []  
    for i in range(15):
        peptide_sam[:,i,:] = Ala
        BA = BAmodel.predict([peptide_sam,MHC_sam])
        importance = pred_ori - IMM([peptide_sam,MHC_sam,BA]).numpy().mean()
        array.append(importance)
        peptide_sam = peptide[index].copy()
    ranking = np.argsort(array) + 1
    tmp = []
    for i in range(15):
        tmp.append(list(ranking).index(i+1))
    position_rank[m,:] = tmp
np.save("/public/slst/home/wanggsh/Data/MHCII_immuno/position_importance/position_ala.npy",position_rank)
```
#### plt figure

```{python, message = FALSE }
#python
from collections import Counter
position_rank = np.load("../data/position.npy")
cmap = mpl.cm.get_cmap('tab20')
delim = np.linspace(0,1,15)
colors = [mpl.colors.rgb2hex(cmap(i)[:4]) for i in delim]
fig,ax = plt.subplots(figsize = (6,4))
for i in np.arange(15):
    y = list(Counter(position_rank[:,i]+1).keys())
    s = list(Counter(position_rank[:,i]+1).values())
    ax.scatter([i for n in range(len(y))],y, s=[m*3.5 for m in s],c=colors[i])
ax.set_ylim(0,16)
ax.set_yticks(np.arange(15)+1)
ax.set_xticks(np.arange(15))
ax.set_xticklabels(['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15'])
ax.set_xlabel('Zero setting')
ax.set_ylabel('Ranking(ascending)')
ax.grid(True,alpha=0.2)
h1 = [ax.plot([],[],color='grey',marker='o',markersize=i,ls='')[0] for i in range(8,15,2)]
leg1 = ax.legend(handles=h1,labels=[10,40,70,100],title='Frequency',loc='lower left',bbox_to_anchor=(0.8,0.6),frameon=False)
ax.add_artist(leg1)
plt.savefig("../figure/position_zero.pdf",dpi = 300,transparent=True)
plt.show()
```

```{python,message = FALSE}
#python
position_rank_ala = np.load("../data/position_ala.npy")
cmap = mpl.cm.get_cmap('tab20')
delim = np.linspace(0,1,15)
colors = [mpl.colors.rgb2hex(cmap(i)[:4]) for i in delim]
fig,ax = plt.subplots(figsize = (6,4))
for i in np.arange(15):
    y = list(Counter(position_rank_ala[:,i]+1).keys())
    s = list(Counter(position_rank_ala[:,i]+1).values())
    ax.scatter([i for n in range(len(y))],y, s=[m*3.5 for m in s],c=colors[i])
ax.set_ylim(0,16)
ax.set_yticks(np.arange(15)+1)
ax.set_xticks(np.arange(15))
ax.set_xticklabels(['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15'])
ax.set_xlabel('Ala scanning')
ax.set_ylabel('Ranking(ascending)')
ax.grid(True,alpha=0.2)
h1 = [ax.plot([],[],color='grey',marker='o',markersize=i,ls='')[0] for i in range(8,15,2)]
leg1 = ax.legend(handles=h1,labels=[10,40,70,100],title='Frequency',loc='lower left',bbox_to_anchor=(0.8,0.6),frameon=False)
ax.add_artist(leg1)
plt.savefig("../figure/position_ala.pdf",dpi = 300,transparent=True)
plt.show()
```

Both results show that P4 (residue 4), P5, and P6 are very important for model prediction. This result is consistent with the results related to MHC class I molecules, and it is reported that these positions are essential for interacting with the TCR. But unlike MHC I molecules, MHC II molecules exhibit a tendency to have less effect on model predictions due to its position closer to the tail, possibly due to its open binding cleft. The results of Ala-scanning are more stable rather than zero-setting, because Ala also has biological functions.

### BA model

We also do the same things by using BA model.

```{python,eval = FALSE}
import pandas as pd
import tensorflow as tf
import sys
sys.path.append("/public/slst/home/wanggsh/Project/MHCII/utils")
from Blosum62 import blosum62
import numpy as np
import random
import matplotlib.pyplot as plt
import matplotlib as mpl
import os
#os.environ["CUDA_VISIBLE_DEVICES"]="-1" 


immuno_data = pd.read_csv("/public/slst/home/wanggsh/Data/MHCII/netMHCpanII4.0_train_data/BA_data/BA_MHC_label.csv")
immuno_data = immuno_data[immuno_data["score"] == 1]
immuno_data["length"] = immuno_data["peptide"].map(len)
#choose 15mer
immuno_data = immuno_data[immuno_data["length"] == 15]
immuno_data = immuno_data.reset_index(drop=True)


immuno_data["pep_blosum"] = immuno_data["peptide"].apply(blosum62,args=(21,))
immuno_data["MHC_blosum"] = immuno_data["sequence"].apply(blosum62,args=(34,))
peptide = np.empty((len(immuno_data),21,21))
for i in range(len(immuno_data)):
    peptide[i] = immuno_data["pep_blosum"][i].reshape((21,21))
#mhc 
MHC = np.empty((len(immuno_data),34,21))
for i in range(len(immuno_data)):
    MHC[i] = immuno_data["MHC_blosum"][i].reshape((34,21))
Ala = list(blosum62("A",1))

BAmodel = tf.keras.models.load_model("/public/slst/home/wanggsh/Project/Saved_model/new_BA_model/model1")

n = 100
position_rank = np.empty([n,15])
for m in range(n):
    index = np.random.choice(np.arange(len(immuno_data)),2000)
    peptide_sam = peptide[index].copy()
    MHC_sam = MHC[index]
    pred_ori = BAmodel([peptide_sam,MHC_sam]).numpy().mean()  
    array = []  
    for i in range(15):
        peptide_sam[:,i,:] = 0
        importance = pred_ori - BAmodel([peptide_sam,MHC_sam]).numpy().mean()
        array.append(importance)
        peptide_sam = peptide[index].copy()
    ranking = np.argsort(array) + 1
    tmp = []
    for i in range(15):
        tmp.append(list(ranking).index(i+1))
    position_rank[m,:] = tmp
np.save("/public/slst/home/wanggsh/Data/MHCII_immuno/position_importance/position_zero_BA.npy",position_rank)

n = 100
position_rank_ala = np.empty([n,15])
for m in range(n):
    index = np.random.choice(np.arange(len(immuno_data)),2000)
    peptide_sam = peptide[index].copy()
    MHC_sam = MHC[index]
    pred_ori = BAmodel([peptide_sam,MHC_sam]).numpy().mean()  
    array = []  
    for i in range(15):
        peptide_sam[:,i,:] = Ala
        importance = pred_ori - BAmodel([peptide_sam,MHC_sam]).numpy().mean()
        array.append(importance)
        peptide_sam = peptide[index].copy()
    ranking = np.argsort(array) + 1
    tmp = []
    for i in range(15):
        tmp.append(list(ranking).index(i+1))
    position_rank[m,:] = tmp
np.save("/public/slst/home/wanggsh/Data/MHCII_immuno/position_importance/position_ala_BA.npy",position_rank)
```

```{python}
#python
from collections import Counter
position_rank = np.load("../data/position_zero_BA.npy")
cmap = mpl.cm.get_cmap('tab20')
delim = np.linspace(0,1,15)
colors = [mpl.colors.rgb2hex(cmap(i)[:4]) for i in delim]
fig,ax = plt.subplots(figsize = (6,6))
for i in np.arange(15):
    y = list(Counter(position_rank[:,i]+1).keys())
    s = list(Counter(position_rank[:,i]+1).values())
    ax.scatter([i for n in range(len(y))],y, s=[m*3.5 for m in s],c=colors[i])
ax.set_ylim(0,16)
ax.set_yticks(np.arange(15)+1)
ax.set_xticks(np.arange(15))
ax.set_xticklabels(['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15'])
ax.set_xlabel('Zero setting')
ax.set_ylabel('Ranking(ascending)')
ax.grid(True,alpha=0.2)
h1 = [ax.plot([],[],color='grey',marker='o',markersize=i,ls='')[0] for i in range(8,15,2)]
leg1 = ax.legend(handles=h1,labels=[10,40,70,100],title='Frequency',loc='lower left',bbox_to_anchor=(0.98,0.2),frameon=False)
ax.add_artist(leg1)
plt.savefig("../figure/position_zero_BA.pdf",dpi = 300,bbox_inches='tight',transparent=True)
plt.show()
```

```{python}
#python
position_rank_ala = np.load("../data/position_ala_BA.npy")
cmap = mpl.cm.get_cmap('tab20')
delim = np.linspace(0,1,15)
colors = [mpl.colors.rgb2hex(cmap(i)[:4]) for i in delim]
fig,ax = plt.subplots(figsize = (6,6))
for i in np.arange(15):
    y = list(Counter(position_rank_ala[:,i]+1).keys())
    s = list(Counter(position_rank_ala[:,i]+1).values())
    ax.scatter([i for n in range(len(y))],y, s=[m*3.5 for m in s],c=colors[i])
ax.set_ylim(0,16)
ax.set_yticks(np.arange(15)+1)
ax.set_xticks(np.arange(15))
ax.set_xticklabels(['1','2','3','4','5','6','7','8','9','10','11','12','13','14','15'])
ax.set_xlabel('Ala scaning')
ax.set_ylabel('Ranking(ascending)')
ax.grid(True,alpha=0.2)
h1 = [ax.plot([],[],color='grey',marker='o',markersize=i,ls='')[0] for i in range(8,15,2)]
leg1 = ax.legend(handles=h1,labels=[10,40,70,100],title='Frequency',loc='lower left',bbox_to_anchor=(0.98,0.2),frameon=False)
ax.add_artist(leg1)
plt.savefig("../figure/position_ala_BA.pdf",dpi = 300,bbox_inches='tight',transparent=True)
plt.show()
```

Furthermore, the BA model does not show the same trend, it is possible that the features learned by the affinity prediction model are the interaction between MHC and peptide compared to the features learned by the immunogenicity prediction model.